From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheKit <nekit1000@gmail.com>
Date: Mon, 30 Aug 2021 00:03:52 +0300
Subject: [PATCH] (halium) disable GWP-ASan and use emulated TLS in libc by
 default

Building locale.cpp with platform TLS and disabled direct TLS slot
usage in earlier patch breaks linker compiled as static binary.

Change-Id: Ic63b814917913ba6055aa9ba2c8e70c098b428af
---
 libc/Android.bp                   |  3 ++-
 libc/bionic/gwp_asan_wrappers.cpp | 21 +++++++++++++++++++++
 2 files changed, 23 insertions(+), 1 deletion(-)

diff --git a/libc/Android.bp b/libc/Android.bp
index e247012ed615214fe6f9758023d3c0f57d4996b4..bbe7b63bd998fe39dd5541334f12bc1fb1630575 100644
--- a/libc/Android.bp
+++ b/libc/Android.bp
@@ -68,7 +68,8 @@ libc_common_flags = [
     "-Wexit-time-destructors",
 
     // GWP-ASan requires platform TLS.
-    "-fno-emulated-tls",
+    // Halium: ELF TLS currently breaks libhybris, so disable it
+    "-femulated-tls",
 
     // We know clang does a lot of harm by rewriting what we've said, and sadly
     // never see any good it does, so let's just ask it to do what we say...
diff --git a/libc/bionic/gwp_asan_wrappers.cpp b/libc/bionic/gwp_asan_wrappers.cpp
index fc59c888207e09fdc4103d934c2187a85756dbdf..777d0ae8894e5fce5b404d2393f408ea48a5e8a8 100644
--- a/libc/bionic/gwp_asan_wrappers.cpp
+++ b/libc/bionic/gwp_asan_wrappers.cpp
@@ -53,6 +53,7 @@
 #include "bionic/malloc_common_dynamic.h"
 #endif  // LIBC_STATIC
 
+#ifdef DISABLED_FOR_HYBRIS_SUPPORT
 static gwp_asan::GuardedPoolAllocator GuardedAlloc;
 static const MallocDispatch* prev_dispatch;
 
@@ -433,3 +434,23 @@ bool EnableGwpAsan(const android_mallopt_gwp_asan_options_t& options) {
       [&](libc_globals* globals) { ret_value = MaybeInitGwpAsan(globals, options); });
   return ret_value;
 }
+
+#else // DISABLED_FOR_HYBRIS_SUPPORT
+
+bool EnableGwpAsan(const android_mallopt_gwp_asan_options_t& /*options*/) {
+  return false;
+}
+
+bool MaybeInitGwpAsanFromLibc(libc_globals* /*globals*/) {
+  return false;
+}
+
+bool MaybeInitGwpAsan(libc_globals* /*globals*/, bool /*force_init*/) {
+  return false;
+}
+
+bool DispatchIsGwpAsan(const MallocDispatch* /*dispatch*/) {
+  return false;
+}
+
+#endif // DISABLED_FOR_HYBRIS_SUPPORT
